x/programs/rust/wasmlanche_sdk/src/state.rs

```diff
@@ -77,3 +78,40 @@ impl State {
         from_slice(&val).map_err(|_| StateError::InvalidBytes)
     }
 }
+
+/// Converts a raw pointer to a deserialized value.
+/// Expects the first 4 bytes of the pointer to represent the [length] of the serialized value,
+/// with the subsequent [length] bytes comprising the serialized data.
+// pub fn from_raw_ptr<V>(ptr: i64) -> V
+// where
+//     V: serde::de::DeserializeOwned,
+// {
+//     let (bytes, _) = bytes_and_length(ptr);
+//     from_slice(&bytes).expect("failed to deserialize")
+// }
+
+// Converts a raw pointer to a type that implements the Argument trait.
+// Expects the first 4 bytes of the pointer to represent the [length] of the serialized value,
+// with the subsequent [length] bytes comprising the serialized data.
+#[must_use]
+pub fn from_raw_ptr<V>(ptr: i64) -> V
+where
+    V: Argument,
+{
+    let (bytes, _) = bytes_and_length(ptr);
+    V::from_bytes(&bytes)
+}
+
+// TODO: move this logic to return a Memory struct that conatins ptr + length
+/// Returns a tuple of the bytes and length of the argument.
+/// # Panics
+/// Panics if the value cannot be converted from i32 to usize.
+/// # Safety
+/// This function is unsafe because it dereferences raw pointers.
```

Comment:
🤔  haha. If the function is `unsafe`, it should be marked as such. 

We can figure out how to fix it later, but it should be glaringly obvious where our unsafe calls are. 

Also, if you look at the [Safety parameters for `from_raw_parts`](https://doc.rust-lang.org/stable/std/slice/fn.from_raw_parts.html#safety), it looks like we have a bunch of these covered because we're using `u8` as `T` (don't have to worry about alignment, etc. 

Would it make sense for the host to call the `program` to do the allocation, guaranteeing the memory exists? Then the guest could place the bytes into the memory location? That's the only thing we're missing for safety (I think).

Reply:
I'm not sure what you mean by 
> The guest could place the bytes into the memory location? 

the host allocated and writes the data for these pointers when it creates parameters.

Reply:
also if I mark this function as unsafe, then all public functions need to be marked unsafe(since the macro calls this function).

Reply:
> I'm not sure what you mean by
> 
> > The guest could place the bytes into the memory location?
> 
> the host allocated and writes the data for these pointers when it creates parameters.

I need to learn more about the different ways arguments can be passed back and forth. You can ignore my question.

Reply:
> also if I mark this function as unsafe, then all public functions need to be marked unsafe(since the macro calls this function).

You can put the function call in an `unsafe` block inside the macro. It's okay though, don't worry about it for now.